contracts/LockProxySwitcheo.scilla

```diff
@@ -0,0 +1,616 @@
+transition lock(
+  tokenAddr: ByStr20,
+  targetProxyHash: ByStr, 
+  toAddress: ByStr,
+  toAssetHash: ByStr,
+  feeAddr: ByStr,
+  amount: Uint256,
+  feeAmount: Uint256,
+  callAmount: Uint256
+)
+  (* todo handle call amount*)
+  amount_uint128_o = builtin to_uint128 amount;
+  call_amount_uint128_o = builtin to_uint128 callAmount;
+  match amount_uint128_o with
+  | Some amount_uint128 =>
+    match call_amount_uint128_o with
+    | Some call_amount_uint128 =>
+      TransferToContract tokenAddr amount_uint128 amount_uint128;
+      lockProcedure tokenAddr targetProxyHash toAssetHash _sender toAddress amount feeAmount feeAddr
+    | None =>
+      e = InvalidCallAmount;
+      ThrowError e
+    end
+  | None =>
+    e = IllegalAmount;
+    ThrowError e
+  end
+end
+
+(* Marks an asset as registered by mapping the asset address to the specified *)
+(* fromContractAddr and assetHash on Swwitcheo TradeHub*)
+(* @param args: the serialized  RegisterAssetTxArgs *)
+(* @param fromContractAddr: the associated contract address on Swticheo TradeHub *)
+(* @param fromChainId: the originating chainId *)
+transition registerAsset(args: ByStr, fromContractAddr: ByStr, fromChainId: Uint64)
+  IsManagerContract _sender;
+  ValidateCounterpartChainId fromChainId;
+  e = { _eventname: "registerAssetInput"; args: args };
+  event e;
+  register_asset_tx_args_o = deserialize_register_asset_tx_args args uint32_zero;
+  match register_asset_tx_args_o with
+  | Some (RegisterAssetTxArgs assetHash nativeAssetHash) =>
+    e = { _eventname: "registerAssetDeserialized"; assetHash: assetHash; nativeAssetHash: nativeAssetHash};
+    event e;
+    nativeAssetAddr_o = builtin to_bystr20 nativeAssetHash;
+    match nativeAssetAddr_o with
+    | Some nativeAssetAddr =>
+      MarkAssetAsRegistered nativeAssetAddr fromContractAddr assetHash;
+      e = { _eventname: "registerAsset"; fromContractAddr: fromContractAddr };
+      event e
+    | None =>
+      e = IllegalRegisterAssetArgs;
+      ThrowError e
+    end
+  | None =>
+    e = DeserializeRegisterAssetArgsFail;
+    ThrowError e
+  end
+end
+
+(* Perfroms a withdrawal that was initiated on Switcheo TradeHub *)
+(* @param argsBz: the serialized TransferTxArgs *)
+(* @param fromContractAddr: the associated contract address on Swticheo TradeHub *)
+(* @param fromChainId: the originating chainId *)
+transition unlock(
+  args: ByStr,
+  fromContractAddr: ByStr,
+  fromChainId: Uint64
+)
+  IsManagerContract _sender;
+  validateFromChainId fromChainId;
+  transfer_tx_args_o = deserialize_transfer_tx_args args uint32_zero;
+  match transfer_tx_args_o with
+  | Some (TransferTxArgs fromAssetHash toAssetHash toAddress fromAddress amount feeAmount feeAddress n) =>
+    ValidataHashLength fromAssetHash;
+    validateAddrLength toAssetHash;
+    tokenAddress_o = builtin to_bystr20 toAssetHash;
+    toAddress_o = builtin to_bystr20 toAddress;
```

Comment:
no need to validateAddrLength toAddress ?

Reply:
added.